modules/frontend/querysharding.go

```diff
@@ -154,25 +160,36 @@ func mergeResponses(ctx context.Context, rrs []RequestResponse) (*http.Response,
 
 	var errCode = http.StatusOK
 	var errBody io.ReadCloser
-	var combinedTrace []byte
-	var shardMissCount = 0
+	var combinedTrace *tempopb.Trace
+	var combinedTraceBytes []byte
+	var shardMissCount, totalBlockErrCount int
 	for _, rr := range rrs {
-		if rr.Response.StatusCode == http.StatusOK {
+		partialContent := rr.Response.StatusCode == http.StatusPartialContent
+		if rr.Response.StatusCode == http.StatusOK || partialContent {
 			body, err := io.ReadAll(rr.Response.Body)
 			rr.Response.Body.Close()
 			if err != nil {
 				return nil, errors.Wrap(err, "error reading response body at query frontend")
 			}
 
-			if len(combinedTrace) == 0 {
-				combinedTrace = body
-			} else {
-				combinedTrace, _, err = model.CombineTraceBytes(combinedTrace, body, model.TracePBEncoding, model.TracePBEncoding)
-				if err != nil {
-					// will result in a 500 internal server error
-					return nil, errors.Wrap(err, "error combining traces at query frontend")
+			var resp tempopb.TraceByIDResponse
+			err = proto.Unmarshal(body, &resp)
+			if err != nil {
+				return nil, errors.Wrap(err, "error reading response body at query frontend")
+			}
+
+			if partialContent {
+				totalBlockErrCount += int(resp.BlockErrCount)
+				if totalBlockErrCount > maxBlockErrCount {
+					return nil, fmt.Errorf("too many block queries failed (max %d)", maxBlockErrCount)
 				}
```

Comment:
I'm wondering if it's useful to fail on `maxBlockErrCount`. We already did the work: all the queriers returned something. Instead of throwing away the results we can still return whatever we got.